File: 138-copy-list-with-random-pointer/copy-list-with-random-pointer.ts
/**
 * Definition for _Node.
 * class _Node {
 *     val: number
 *     next: _Node | null
 *     random: _Node | null
 * 
 *     constructor(val?: number, next?: _Node, random?: _Node) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *         this.random = (random===undefined ? null : random)
 *     }
 * }
 */


function copyRandomList(head: _Node | null): _Node | null {
    if (!head) return null

    let current: _Node | null = head

    while (current !== null) {
        const newNode: _Node = new _Node(current.val, current.next, null)
        current.next = newNode
        current = newNode.next
    }

    current = head

    while (current !== null) {
        if (current.random !== null) {
            current.next.random = current.random.next
        }
        current = current.next.next
    }

    current = head
    const copyHead: _Node | null = head.next
    let copyCurrent: _Node | null = copyHead

    while (current !== null) {
        current.next = current.next.next
        if (copyCurrent.next !== null) {
            copyCurrent.next = copyCurrent.next.next
        }

        current = current.next
        copyCurrent = copyCurrent.next
    }
    return copyHead

};

File: 138-copy-list-with-random-pointer/copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        old_to_new = defaultdict(lambda: None)
        curr = head

        while curr:
            old_to_new[curr] = Node(curr.val)
            curr = curr.next

        curr = head

        while curr:
            new_node = old_to_new[curr]
            if curr.next:
                new_node.next = old_to_new[curr.next]
            if curr.random:
                new_node.random = old_to_new[curr.random]
            curr = curr.next

        return old_to_new[head]